### DISCUSSION TITLE
Is there type-matching completion in clangd?

### DISCUSSION THREAD

**Discussion by podtserkovskiy** [clangd/clangd#1143] [Q&A] [answered]
Hey, 

I'm trying to use clangd with vscode via [vscode-clangd](https://github.com/clangd/vscode-clangd) extension. 
When I'm triggering completion on function args or assignments I see lots of possible values despite their types without any sensible ordering. 
For example:
- for `void func1(bool val)` I expect true/false literals and boolean variables
- for `void func2(SomeEnum val)` I expect SomeEnumtype values
- when I assign a value to `bool var1;` I expect true/false literals and boolean variables

I'm not sure if it's a bug/missing feature in clangd/vscode-clangd/vscode or incorrect settings in my environment. 
Do you guys know any details about it? 

https://user-images.githubusercontent.com/12721466/167259860-08f2754d-ac00-40fa-844d-99da40cf7ce6.mov

**Comment by HighCommander4** [comment, marked as answer]
There is definitely some logic in clangd for ranking completion proposals of the relevant type higher (e.g. the logic added in https://reviews.llvm.org/D52276).

I don't think there's any logic for boolean (or other) literals, which is why you're not seeing `true` or `false`.

In the video, the main thing that seems unexpected is that when completing the argument of `bb()`, the local variables of type `const char*` are ranked higher than the enumerators of the expected type `Alphabet`, that's worth investigating further.

**Reply by podtserkovskiy** [reply]
Thanks for the answer, that looks interesting, I'll try to experiment with this more on a real-world project 😀
